boot/bootstate20.go

```diff
@@ -525,6 +538,52 @@ func (ks20 *bootState20Kernel) commit() error {
 	return nil
 }
 
+// selectAndCommitSnapInitramfsMount chooses which snap should be mounted
+// during the initramfs, and commits that choice if it needs state updated.
+// Choosing to boot/mount the base snap needs to be committed to the
+// modeenv, but no state needs to be committed when choosing to mount a
+// kernel snap.
+func (ks20 *bootState20Kernel) selectAndCommitSnapInitramfsMount() (sn snap.PlaceInfo, err error) {
+	err = ks20.kModeenv.loadModeenv()
+	if err != nil {
+		return nil, err
+	}
+
+	// first do the generic choice of which snap to use
+	first, second, expectFallback, err := genericInitramfsSelectSnap(ks20, TryingStatus, "kernel")
+	if err != nil {
+		return nil, err
+	}
+
+	// now validate the chosen kernel snap against the modeenv CurrentKernel's
+	// setting
+	// make a map to easily check if a kernel snap is valid or not
+	validKernels := make(map[string]bool, len(ks20.kModeenv.modeenv.CurrentKernels))
+	for _, validKernel := range ks20.kModeenv.modeenv.CurrentKernels {
+		validKernels[validKernel] = true
+	}
+
+	// always try the first and fallback to the second if we fail
+	if validKernels[first.Filename()] {
```

Comment:
could we use strutil.ListContain instead? This is going to be a couple of snaps at most

Reply:
sure